File: go/streaming/internal/stream_asr_job_outgoing_message/submit_audio_fragment_message.go
package stream_asr_job_outgoing_message

import "encoding/base64"

type SubmitAudioFragmentMessage struct {
	Command string                         `json:"command"`
	Body    SubmitAudioFragmentMessageBody `json:"body"`
}

type SubmitAudioFragmentMessageBody struct {
	ChannelIndex       int    `json:"channel_index"`
	AudioFragmentIndex int    `json:"audio_fragment_index"`
	Base64AudioData    string `json:"base64_audio_data"`
}

func NewSubmitAudioFragmentMessage(
	channelIndex int,
	audioFragmentIndex int,
	audioData []byte,
) SubmitAudioFragmentMessage {
	return SubmitAudioFragmentMessage{
		Command: "submit_audio_fragment",
		Body: SubmitAudioFragmentMessageBody{
			ChannelIndex:       channelIndex,
			AudioFragmentIndex: audioFragmentIndex,
			Base64AudioData:    base64.StdEncoding.EncodeToString(audioData),
		},
	}
}

File: cpp/src/poetics_sdk/streaming/asr_job/outgoing_message/submit_audio_fragment_message.cpp
#include "submit_audio_fragment_message.h"

namespace poetics::streaming::asr_job::outgoing_message {
    SubmitAudioFragmentMessageBody::SubmitAudioFragmentMessageBody()
    {
    }

    SubmitAudioFragmentMessageBody::SubmitAudioFragmentMessageBody(int _channelIndex, int _audioFragmentIndex, string _base64AudioData) 
        : channelIndex(_channelIndex), audioFragmentIndex(_audioFragmentIndex), base64AudioData(_base64AudioData)
    {
    }

    SubmitAudioFragmentMessage::SubmitAudioFragmentMessage(string _command, SubmitAudioFragmentMessageBody _body)
    {
        command = _command;
        body = _body;
    }

    void to_json(json &j, const SubmitAudioFragmentMessageBody &body)
    {
        j["channel_index"] = body.channelIndex;
        j["audio_fragment_index"] = body.audioFragmentIndex;
        j["base64_audio_data"] = body.base64AudioData;
    }

    void to_json(json &j, const SubmitAudioFragmentMessage &message)
    {
        j = json{
            {"command", message.command},
            {"body", message.body}
        };
    }
}
